[macOS] Network connection logs should be silenced in Safari private browsing mode

whsieh opened the pull request:
#### 01329a31ae593125485a4f2bcac7d76981fbcf2f
<pre>
[macOS] Network connection logs should be silenced in Safari private browsing mode
<a href="https://bugs.webkit.org/show_bug.cgi?id=259022">https://bugs.webkit.org/show_bug.cgi?id=259022</a>
rdar://111586011

Reviewed by Youenn Fablet.

In Safari (and other browsers) on iOS, we avoid logging any information about network connections —
even with redaction — in private browsing (i.e. when using an ephemeral data store). We achieve this
by passing in `nw_context_privacy_level_silent` when configuring `NSURLSessionConfiguration` in the
network process.

However, this currently doesn&apos;t work on macOS because the logic for determining whether or not the
parent of the networking process is a full web browser (`isParentProcessAFullWebBrowser()`) only
takes the `com.apple.developer.web-browser` entitlement into account, which is iOS-specific.

In contrast, `isFullWebBrowserOrRunningTest()` (also implemented in `DefaultWebBrowserChecks.mm`)
uses the above entitlement check on iOS, and checks whether the application can open HTTP-family
URLs on macOS. Since this function is intended to be invoked from the application process, we add a
flag to `NetworkProcessCreationParameters` to compute and send this over to the network process upon
initialization.

Test: WKWebsiteDataStore.DoNotLogNetworkConnectionsInEphemeralSessions

* Source/WebKit/NetworkProcess/NetworkProcess.h:
(WebKit::NetworkProcess::isParentProcessFullWebBrowserOrRunningTest const):
* Source/WebKit/NetworkProcess/NetworkProcessCreationParameters.h:
* Source/WebKit/NetworkProcess/NetworkProcessCreationParameters.serialization.in:

See above for more details.

* Source/WebKit/NetworkProcess/cocoa/NetworkProcessCocoa.mm:
(WebKit::NetworkProcess::platformInitializeNetworkProcessCocoa):
* Source/WebKit/NetworkProcess/cocoa/NetworkSessionCocoa.mm:
(WebKit::SessionWrapper::initialize):

Implement the main fix here — replace `isParentProcessAFullWebBrowser(networkProcess)` with
`networkProcess.isParentProcessFullWebBrowserOrRunningTest()`.

(WebKit::NetworkSessionCocoa::NetworkSessionCocoa):
(WebKit::NetworkSessionCocoa::sessionWrapperForTask):
(WebKit::NetworkSessionCocoa::removeNetworkWebsiteData):

Add a few FIXMEs referencing the fact that the other existing `isParentProcessAFullWebBrowser`
checks in this file don&apos;t work on macOS.

* Source/WebKit/UIProcess/Network/NetworkProcessProxy.cpp:
(WebKit::NetworkProcessProxy::sendCreationParametersToNewProcess):
* Tools/TestWebKitAPI/Tests/WebKit/AdvancedPrivacyProtections.mm:
(-[TestWKWebView collectLogsForNewConnections]): Deleted.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebsiteDatastore.mm:

Add an API test to verify that no network connection logging is saved when navigating in an
ephemeral session.

* Tools/TestWebKitAPI/cocoa/TestWKWebView.h:
* Tools/TestWebKitAPI/cocoa/TestWKWebView.mm:
(-[TestWKWebView collectLogsForNewConnections]):

Move this testing helper into `TestWKWebView`, so that we can invoke it from API tests across
different files.

Canonical link: <a href="https://commits.webkit.org/265904@main">https://commits.webkit.org/265904@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f3efbb58c0755981bb98e533d8e285b19e8c74ad

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/12151 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/12498 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/12800 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/13897 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/11726 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/12182 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/14913 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/12512 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/14410 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/12316 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/13121 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/10290 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/14313 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/10407 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/11031 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/18146 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/11488 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/11190 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/14366 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/11681 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/9630 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/10897 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/2999 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/15223 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/11534 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

youennf commented on Source/WebKit/NetworkProcess/cocoa/NetworkSessionCocoa.mm:
```diff
@@ -1568,6 +1569,7 @@ static void activateSessionCleanup(NetworkSessionCocoa& session, const NetworkSe
 SessionWrapper& NetworkSessionCocoa::sessionWrapperForTask(WebPageProxyIdentifier webPageProxyID, const WebCore::ResourceRequest& request, WebCore::StoredCredentialsPolicy storedCredentialsPolicy, std::optional<NavigatingToAppBoundDomain> isNavigatingToAppBoundDomain)
 {
     auto shouldBeConsideredAppBound = isNavigatingToAppBoundDomain ? *isNavigatingToAppBoundDomain : NavigatingToAppBoundDomain::Yes;
+    // FIXME: The following `isParentProcessAFullWebBrowser` check is inaccurate in Safari on macOS.
```
Should we specialise isParentProcessAFullWebBrowser for NetworkProcess&?
Or move the boolean flag from NetworkProcess to AuxiliaryProcess and check it in isParentProcessAFullWebBrowser.

whsieh replied:
Indeed — I was thinking of doing the latter (so that the "web browser" checks in all of the WebKit child processes would be accurate, not just for this case).

However, to limit risk for now, I'll do that as a followup patch in https://webkit.org/b/259056 (which should hopefully fix other similar bugs in macOS Safari).

Thanks for the review!

whsieh committed:
[macOS] Network connection logs should be silenced in Safari private browsing mode
https://bugs.webkit.org/show_bug.cgi?id=259022
rdar://111586011

Reviewed by Youenn Fablet.

In Safari (and other browsers) on iOS, we avoid logging any information about network connections —
even with redaction — in private browsing (i.e. when using an ephemeral data store). We achieve this
by passing in `nw_context_privacy_level_silent` when configuring `NSURLSessionConfiguration` in the
network process.

However, this currently doesn't work on macOS because the logic for determining whether or not the
parent of the networking process is a full web browser (`isParentProcessAFullWebBrowser()`) only
takes the `com.apple.developer.web-browser` entitlement into account, which is iOS-specific.

In contrast, `isFullWebBrowserOrRunningTest()` (also implemented in `DefaultWebBrowserChecks.mm`)
uses the above entitlement check on iOS, and checks whether the application can open HTTP-family
URLs on macOS. Since this function is intended to be invoked from the application process, we add a
flag to `NetworkProcessCreationParameters` to compute and send this over to the network process upon
initialization.

Test: WKWebsiteDataStore.DoNotLogNetworkConnectionsInEphemeralSessions

* Source/WebKit/NetworkProcess/NetworkProcess.h:
(WebKit::NetworkProcess::isParentProcessFullWebBrowserOrRunningTest const):
* Source/WebKit/NetworkProcess/NetworkProcessCreationParameters.h:
* Source/WebKit/NetworkProcess/NetworkProcessCreationParameters.serialization.in:

See above for more details.

* Source/WebKit/NetworkProcess/cocoa/NetworkProcessCocoa.mm:
(WebKit::NetworkProcess::platformInitializeNetworkProcessCocoa):
* Source/WebKit/NetworkProcess/cocoa/NetworkSessionCocoa.mm:
(WebKit::SessionWrapper::initialize):

Implement the main fix here — replace `isParentProcessAFullWebBrowser(networkProcess)` with
`networkProcess.isParentProcessFullWebBrowserOrRunningTest()`.

(WebKit::NetworkSessionCocoa::NetworkSessionCocoa):
(WebKit::NetworkSessionCocoa::sessionWrapperForTask):
(WebKit::NetworkSessionCocoa::removeNetworkWebsiteData):

Add a few FIXMEs referencing the fact that the other existing `isParentProcessAFullWebBrowser`
checks in this file don't work on macOS.

* Source/WebKit/UIProcess/Network/NetworkProcessProxy.cpp:
(WebKit::NetworkProcessProxy::sendCreationParametersToNewProcess):
* Tools/TestWebKitAPI/Tests/WebKit/AdvancedPrivacyProtections.mm:
(-[TestWKWebView collectLogsForNewConnections]): Deleted.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebsiteDatastore.mm:

Add an API test to verify that no network connection logging is saved when navigating in an
ephemeral session.

* Tools/TestWebKitAPI/cocoa/TestWKWebView.h:
* Tools/TestWebKitAPI/cocoa/TestWKWebView.mm:
(-[TestWKWebView collectLogsForNewConnections]):

Move this testing helper into `TestWKWebView`, so that we can invoke it from API tests across
different files.

Canonical link: https://commits.webkit.org/265904@main
